### PR TITLE
[SYCL] Add OSUtil::getCurrentDSODir() function

### PR DESCRIPTION
**Pull request by asavonic**
The function returns a directory where the current dynamic shared
object (libsycl.so or sycl.dll in our case) was found.

This can be useful to access files relative to the libsycl.so
directory: config files, supplemental libraries, etc.

Signed-off-by: Andrew Savonichev <andrew.savonichev@intel.com>

### CODE REVIEW

File: sycl/source/detail/os_util.cpp [hunk 1]
```diff
@@ -77,6 +87,97 @@ OSModuleHandle OSUtil::getOSModuleHandle(const void *VirtAddr) {
...
+  //
+  //  3) Skip offset, dev, inode
+  //
+  //  4) Extract an absolute path to a filename and get a dirname from it.
+  //
+  uintptr_t CurrentFunc = (uintptr_t) &getCurrentDSODir;
+  std::ifstream Stream("/proc/self/maps");
+  Stream >> std::hex;
+  while (!Stream.eof()) {
+    if (procMapsAddressInRange(Stream, CurrentFunc)) {
```
Comment (AGindinson) [inline sycl/source/detail/os_util.cpp]: Would switching to an early `continue` be better? Taking into account line 168, of course

Comment (asavonic) [reply sycl/source/detail/os_util.cpp]: Sure, done.

File: sycl/source/detail/os_util.cpp [hunk 2]
```diff
@@ -77,6 +87,99 @@ OSModuleHandle OSUtil::getOSModuleHandle(const void *VirtAddr) {
...
+  assert(!Stream.fail() && Stream.peek() == ' ' &&
+         "Couldn't read /proc/self/maps correctly");
+  Stream.ignore(1);
+
+  return Addr >= Start && Addr < End;
+}
+
+/// Returns an absolute path to a directory where the object was found.
+std::string OSUtil::getCurrentDSODir() {
+  // Examine /proc/self/maps and find where ^~~this function comes from - this
```
Comment (romanovvlad) [inline sycl/source/detail/os_util.cpp]: ^~~this is confusing, could you please just name the function?

File: sycl/source/detail/os_util.cpp [hunk 3]
```diff
@@ -77,6 +87,99 @@ OSModuleHandle OSUtil::getOSModuleHandle(const void *VirtAddr) {
   return reinterpret_cast<OSModuleHandle>(Res.Handle);
 }
 
+bool procMapsAddressInRange(std::istream &Stream, uintptr_t Addr) {
+  uintptr_t Start, End;
```
Comment (romanovvlad) [inline sycl/source/detail/os_util.cpp]:
```suggestion
  uintptr_t Start = 0, End = 0;
```

File: sycl/source/detail/os_util.cpp [hunk 4]
```diff
@@ -77,6 +87,99 @@ OSModuleHandle OSUtil::getOSModuleHandle(const void *VirtAddr) {
...
+  std::ifstream Stream("/proc/self/maps");
+  Stream >> std::hex;
+  while (!Stream.eof()) {
+    if (!procMapsAddressInRange(Stream, CurrentFunc)) {
+      // Skip the rest until an EOL and check the next line
+      Stream.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+      continue;
+    }
+
+    char Perm[4];
```
Comment (romanovvlad) [inline sycl/source/detail/os_util.cpp]:
```suggestion
    char Perm[4] = {0};
```

File: sycl/source/detail/os_util.cpp [hunk 5]
```diff
@@ -77,6 +87,99 @@ OSModuleHandle OSUtil::getOSModuleHandle(const void *VirtAddr) {
...
+    Stream.readsome(Perm, sizeof(Perm));
+    assert(Perm[0] == 'r' && Perm[2] == 'x' &&
+           "Invalid flags in /proc/self/maps");
+    assert(Stream.peek() == ' ');
+    Stream.ignore(1);
+
+    // Read and ignore the following:
+    uintptr_t Offset, DevMajor, DevMinor, Inode;
```
Comment (romanovvlad) [inline sycl/source/detail/os_util.cpp]: Can we avoid unnecessary reading and str2int conversion by just calling  "ignore" several times with space as a delimiter?

File: sycl/source/detail/os_util.cpp [hunk 6]
```diff
@@ -77,6 +87,99 @@ OSModuleHandle OSUtil::getOSModuleHandle(const void *VirtAddr) {
...
+    Stream.getline(Path, PATH_MAX - 1);
+    Path[PATH_MAX - 1] = '\0';
+    return OSUtil::getDirName(Path);
+  }
+  assert(false && "Unable to find the current function in /proc/self/maps");
+  return "";
+}
+
+std::string OSUtil::getDirName(const char* Path) {
+  std::string Tmp(Path);
```
Comment (romanovvlad) [inline sycl/source/detail/os_util.cpp]: Probably I'm missing something, but this function does two copies of a string.
Wouldn't it be better to do something like:
```

const char *LastSlash = nullptr;
while(*Path != '\0')
  if(*Path == '/')
   LastSlash = Path;
return std::string(Path, size_t(LastSlash - Path));
```

?

File: sycl/source/detail/os_util.cpp [hunk 7]
```diff
@@ -77,6 +87,97 @@ OSModuleHandle OSUtil::getOSModuleHandle(const void *VirtAddr) {
...
+  // dirname(3) needs a writable C string: a null-terminator is written where a
+  // path should split.
+  size_t TruncatedSize = strlen(dirname(const_cast<char*>(Tmp.c_str())));
```
Comment (romanovvlad) [inline sycl/source/detail/os_util.cpp]:
```suggestion
  size_t TruncatedSize = strlen(dirname(Tmp.data()));
```

File: sycl/source/detail/os_util.cpp [hunk 8]
```diff
@@ -9,8 +9,12 @@
 #include <CL/sycl/detail/os_util.hpp>
 #include <CL/sycl/exception.hpp>
 
+#include <cassert>
+
 #ifdef SYCL_RT_OS_POSIX_SUPPORT
 #include <cstdlib>
+#include <cstring>
+#include <fstream>
```
[the remark this answers is not in the document]
Comment (asavonic) [reply sycl/source/detail/os_util.cpp]: > Why these headers are under `SYCL_RT_OS_POSIX_SUPPORT`? As I can see [here](https://en.cppreference.com/w/cpp/header), all of them are pretty standard C++ headers

They are not used on Windows. Moved under `#if defined(SYCL_RT_OS_LINUX)`

File: sycl/source/detail/os_util.cpp [hunk 9]
```diff
@@ -20,13 +24,19 @@
 #endif // _GNU_SOURCE
 
 #include <cstdio>
+#include <dlfcn.h>
+#include <libgen.h> // for dirname
+#include <link.h>
```
[the remark this answers is not in the document]
Comment (asavonic) [reply sycl/source/detail/os_util.cpp]: Removed.

File: sycl/source/detail/os_util.cpp [hunk 10]
```diff
@@ -20,13 +24,19 @@
...
 #include <link.h>
+#include <linux/limits.h> // for PATH_MAX
+#include <stdio.h>
```
[the remark this answers is not in the document]
Comment (asavonic) [reply sycl/source/detail/os_util.cpp]: Removed.

File: sycl/source/detail/os_util.cpp [hunk 11]
```diff
@@ -93,6 +194,25 @@ OSModuleHandle OSUtil::getOSModuleHandle(const void *VirtAddr) {
   return reinterpret_cast<OSModuleHandle>(PhModule);
 }
 
+/// Returns an absolute path where the object was found.
+std::string OSUtil::getCurrentDSODir() {
+  char Path[MAX_PATH];
+  Path[0] = '\0';
+  Path[sizeof(Path) - 1] = '\0';
```
[the remark this answers is not in the document]
Comment (asavonic) [reply sycl/source/detail/os_util.cpp]: I don't want to initialize the entire array: it is pretty large and the following functions are going to overwrite it.
The first and the last elements should be initialized though, so we don't read garbage if system functions fail and asserts are disabled.

File: sycl/source/detail/os_util.cpp (hunk 7 above)
Comment (asavonic) [reply sycl/source/detail/os_util.cpp]: The string passed to `dirname` should be null-terminated, and AFAIR `std::string::data` does not guarantee this.

Comment (romanovvlad) [reply sycl/source/detail/os_util.cpp]: This page says data and c_str returns the same value starting from C++11:
http://www.cplusplus.com/reference/string/string/data/

File: sycl/source/detail/os_util.cpp (hunk 4 above)
Comment (romanovvlad) [reply sycl/source/detail/os_util.cpp]: OK.

File: sycl/source/detail/os_util.cpp (hunk 7 above)
Comment (asavonic) [reply sycl/source/detail/os_util.cpp]: Anyway, it still returns a `const char*`.

Comment (romanovvlad) [reply sycl/source/detail/os_util.cpp]: Right.

**Comment by romanovvlad** [comment]
@AlexeySachkov Please, "Resolve conversation" if your comments are addressed.

File: sycl/unittests/CMakeLists.txt [hunk 12]
```diff
@@ -9,10 +9,16 @@ function(add_sycl_unittest test_dirname)
   # Enable exception handling for these unit tests
   set(LLVM_REQUIRES_EH 1)
 
+  if (CMAKE_BUILD_TYPE MATCHES "Debug")
+    set(sycl_lib "sycld")
+  else()
+    set(sycl_lib "sycl")
+  endif()
+
```
Comment (romanovvlad) [inline sycl/unittests/CMakeLists.txt]: What is the purpose of this change? Why can't we always use "sycl"?

Comment (asavonic) [reply sycl/unittests/CMakeLists.txt]: Debug tests should link with debug SYCL runtime on Windows, because Microsoft C and C++ runtime are have different ABI in for debug and release.

I made this change because SYCL unittests failed to compile on Windows in Debug build. I can make a separate PR with this change.

Anyway, I forgot to guard this change with `if (MSVC)`, so please don't merge this PR yet.
